sgx/common/src/ec_key_cert_marshal.c

```diff
@@ -0,0 +1,112 @@
+/**
+ * @file ec_key_cert_marshal.c
+ *
+ * @brief Provides implementation of functionality used to marshal (serialize)
+ *        elliptic curve key and cert structs. This supports passing these
+ *        variables across the enclave boundary, as a binary data format is
+ *        required.
+ */
+
+#include "ec_key_cert_marshal.h"
+
+/*****************************************************************************
+ * marshal_ec_pkey_to_der()
+ ****************************************************************************/
+int marshal_ec_pkey_to_der(EVP_PKEY ** ec_pkey_in,
+                           unsigned char **ec_der_bytes_out,
+                           int *ec_der_bytes_out_len)
+{
+  // validate that key to be marshalled is elliptic curve (EC) type
+  EVP_PKEY *pkey_ptr = *ec_pkey_in;
+
+  if (EVP_PKEY_base_id(pkey_ptr) != EVP_PKEY_EC)
+  {
+    const char *src_file_name = __FILE__;
+    const char *func_name = __func__;
+    int severity = 3;
+    const char *message = "PKEY to be marshalled is not of EC type";
+    const int line = __LINE__ + 1;
+    log_event_ocall(&src_file_name, &func_name, &line, &severity, &message);
+    return EXIT_FAILURE;
+  }
+
+  // Validate that a pointer to a NULL buffer pointer was passed in for
+  // the binary output byte array. Put it in that state if necessary.
+  // This will support memory allocation by the call to i2d_PrivateKey()
+  // for PKEY to DER format conversion rather than having to do the
+  // required size calculation and memory allocation beforehand.
+  if (*ec_der_bytes_out != NULL)
+  {
+    free(*ec_der_bytes_out);
```

Comment:
I'd prefer to require that ```*ec_der_bytes_out``` be ```NULL``` and have the function error if it's not. As written this could segfault if the caller passes an uninitialized pointer.

Reply:
Yes, I originally did it that way and then thought that I could make it work even if the caller passes in an already allocated pointer. If an uninitialized pointer is passed in, though, this is still problematic. I replaced the free and assign to NULL statements with return failure. I removed the log message because when I added log messages to the unmarshal code (currently used by trusted code), I realized that even though log_event_ocall() is is known within trusted and untrusted code, the include path is dependent upon where you call it.

Reply:
I mentioned in the previous comment that implementing logging in the "common" functions was problematic because for the trusted code, the declaration of log_ocall() is via the EDGR8R generated header file and for the untrusted code via the log_ocall.h header file directly. In an effort to address this, I renamed the "enclave_log()" macro to "kmyth_sgx_log()" and moved it to a new kmyth_enclave_common.h header file. I added a definition of _KMYTH_LOCALE_TRUSTED_ kmyth_enclave_trusted.h. and added conditional includes in kmyth_enclave_common.h based on the value of this variable. My thought was that if all  "common" functionality is included by kmyth_enclave_trusted.h including kmyth_enclave_common.h, that this would be fairly clean, but this is still a little messy and may or may not be the right approach in the big picture. I also added a temporary test block in the demo app to show that using kmyth_sgx_log() from the untrusted side works.